### ISSUE TITLE
std.trace should not hardcode stderr

### ISSUE THREAD

**Issue by sbarzowski** [google/go-jsonnet#394]
When library interface is used we should not assume that `stderr` is the right target for `std.trace` messages. Instead we should provide some API to define where it should go. Probably a Writer interface here in go-jsonnet, but I'm not sure about C api.

**Comment by sparkprime** [comment]
In C it would need to be a callback like the others, which would be called with the string as a param.  It should be fairly easy compared to some of the other C/go interfaces!

**Comment by hanyucui** [comment]
@sbarzowski Want to give this a try. Hopefully, I can start in a week or two. I imagine I will need to first learn the existing library interfaces. Let me know if you have any advice.

**Comment by sbarzowski** [comment]
Cool!

We can have an io.Writer in VM, which can be set with some method. This can then be used in the Fprintf instead of os.Stderr in std.trace implementation (in builtins.go). That should be enough to support it with Go API.

To support it in C API (which can be a separate PR, and doesn't necessarily need to be done at the same time), we need to figure out the interface for the callback or something – we don't have a ready io.Writer in C. 

One way is to have a callback function which basically mirrors io.Writer and then have a custom loop to output the data through such callback. We can still use sprintf to prepare the message. Maybe you can come up with something better.

**Comment by hanyucui** [comment]
@sbarzowski For the C API, it seems the suggestion from @sparkprime is to pass the message string to C in a callback rather than accepting a `FILE*` and writing to it directly, right?

**Comment by sbarzowski** [comment]
Yeah, that would be better. The reason is that we're trying to make the API OS agnostic. AFAIK it's not possible to pass your custom thing as a FILE* in C. So with `FILE*` t would need to refer to some actual file descriptor. Please, correct me if I'm wrong. Go `Writer` doesn't have this problem – it can be anything consuming bytes.

**Comment by hanyucui** [comment]
Makes sense. Let me give it a try.
